### ISSUE TITLE
Permission denied @ rb_file_s_stat after RubyGems 3.0.0 upgrade

### ISSUE THREAD

**Issue by andrewkendall** [Shopify/bootsnap#220]
I'm tracking an error which seems to be related to an interaction between RubyGems and Bootsnap.

I have multiple systems all running perfectly on RubyGems 2.7.8, after running `gem update --system` to RubyGems 3.0.0 the deployments fail with the following Passenger error:

```
Permission denied @ rb_file_s_stat - /usr/local/rvm/rubies/ruby-2.6.0-rc2/lib/ruby/gems/2.6.0/gems/bundler-1.17.2/lib (Errno::EACCES)
  /usr/local/rvm/gems/ruby-2.6.0-rc2/gems/bootsnap-1.3.2/lib/bootsnap/load_path_cache/path.rb:28:in `stat'
  /usr/local/rvm/gems/ruby-2.6.0-rc2/gems/bootsnap-1.3.2/lib/bootsnap/load_path_cache/path.rb:28:in `non_directory?'
  /usr/local/rvm/gems/ruby-2.6.0-rc2/gems/bootsnap-1.3.2/lib/bootsnap/load_path_cache/cache.rb:134:in `block (2 levels) in push_paths_locked'
  /usr/local/rvm/gems/ruby-2.6.0-rc2/gems/bootsnap-1.3.2/lib/bootsnap/load_path_cache/cache.rb:131:in `each'
  /usr/local/rvm/gems/ruby-2.6.0-rc2/gems/bootsnap-1.3.2/lib/bootsnap/load_path_cache/cache.rb:131:in `block in push_paths_locked'
  /usr/local/rvm/gems/ruby-2.6.0-rc2/gems/bootsnap-1.3.2/lib/bootsnap/load_path_cache/store.rb:45:in `transaction'
  /usr/local/rvm/gems/ruby-2.6.0-rc2/gems/bootsnap-1.3.2/lib/bootsnap/load_path_cache/cache.rb:130:in `push_paths_locked'
  /usr/local/rvm/gems/ruby-2.6.0-rc2/gems/bootsnap-1.3.2/lib/bootsnap/load_path_cache/cache.rb:113:in `block in reinitialize'
  /usr/local/rvm/gems/ruby-2.6.0-rc2/gems/bootsnap-1.3.2/lib/bootsnap/load_path_cache/cache.rb:107:in `synchronize'
  /usr/local/rvm/gems/ruby-2.6.0-rc2/gems/bootsnap-1.3.2/lib/bootsnap/load_path_cache/cache.rb:107:in `reinitialize'
  /usr/local/rvm/gems/ruby-2.6.0-rc2/gems/bootsnap-1.3.2/lib/bootsnap/load_path_cache/cache.rb:14:in `initialize'
  /usr/local/rvm/gems/ruby-2.6.0-rc2/gems/bootsnap-1.3.2/lib/bootsnap/load_path_cache.rb:33:in `new'
  /usr/local/rvm/gems/ruby-2.6.0-rc2/gems/bootsnap-1.3.2/lib/bootsnap/load_path_cache.rb:33:in `setup'
  /usr/local/rvm/gems/ruby-2.6.0-rc2/gems/bootsnap-1.3.2/lib/bootsnap.rb:24:in `setup'
  /usr/local/rvm/gems/ruby-2.6.0-rc2/gems/bootsnap-1.3.2/lib/bootsnap/setup.rb:33:in `<top (required)>'
  /home/application_name/config/boot.rb:4:in `require'
  /home/application_name/config/boot.rb:4:in `<top (required)>'
  /home/application_name/config/application.rb:1:in `require_relative'
  /home/application_name/config/application.rb:1:in `<top (required)>'
  /home/application_name/config/environment.rb:2:in `require_relative'
  /home/application_name/config/environment.rb:2:in `<top (required)>'
  config.ru:3:in `require_relative'
  config.ru:3:in `block in <main>'
  /usr/local/rvm/gems/ruby-2.6.0-rc2/gems/rack-2.0.6/lib/rack/builder.rb:55:in `instance_eval'
  /usr/local/rvm/gems/ruby-2.6.0-rc2/gems/rack-2.0.6/lib/rack/builder.rb:55:in `initialize'
  config.ru:1:in `new'
  config.ru:1:in `<main>'
  /opt/passenger-6.0.0/src/helper-scripts/rack-preloader.rb:101:in `eval'
  /opt/passenger-6.0.0/src/helper-scripts/rack-preloader.rb:101:in `preload_app'
  /opt/passenger-6.0.0/src/helper-scripts/rack-preloader.rb:189:in `block in <module:App>'
  /opt/passenger-6.0.0/src/ruby_supportlib/phusion_passenger/loader_shared_helpers.rb:380:in `run_block_and_record_step_progress'
  /opt/passenger-6.0.0/src/helper-scripts/rack-preloader.rb:188:in `<module:App>'
  /opt/passenger-6.0.0/src/helper-scripts/rack-preloader.rb:30:in `<module:PhusionPassenger>'
  /opt/passenger-6.0.0/src/helper-scripts/rack-preloader.rb:29:in `<main>'
```

Disabling Bootsnap allows the application to boot again.

I'm seeing this in mixed environments including Fedora 28 and CentOS 7.6, Ruby 2.5.1, 2.5.3 and 2.6.0-rc2.

Is anyone else seeing similar issues?

**Comment by benton** [comment]
Yes, this is reproducible.

**Comment by andrewkendall** [comment]
Thank you for confirming @benton, as disabling Bootsnap fixes the issue I'll do that for now until there is more information on the cause. I saw RubyGems has several recently opened issues related to permissions, so possibly this is something that will be fixed on their next updates.

**Comment by andrewkendall** [comment]
Solved with RubyGems 3.0.1.
